core/src/main/java/org/apache/accumulo/core/logging/ConditionalLogger.java

```diff
@@ -0,0 +1,197 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.logging;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import org.slf4j.Logger;
+import org.slf4j.Marker;
+import org.slf4j.event.Level;
+import org.slf4j.helpers.AbstractLogger;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+/**
+ * Logger that wraps another Logger and only emits a log message once per the supplied duration.
+ *
+ */
+public abstract class ConditionalLogger extends AbstractLogger {
+
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * A Logger implementation that will log a message at the supplied elevated level if it has not
+   * been seen in the supplied duration. For repeat occurrences the message will be logged at the
+   * level used in code (which is likely a lower level). Note that the first log message will be
+   * logged at the elevated level because it has not been seen before.
+   */
+  public static class EscalatingLogger extends DeduplicatingLogger {
+
+    private static final long serialVersionUID = 1L;
+    private final Level elevatedLevel;
+
+    public EscalatingLogger(Logger log, Duration threshold, Level elevatedLevel) {
+      super(log, threshold);
+      this.elevatedLevel = elevatedLevel;
+    }
+
+    @Override
+    protected void handleNormalizedLoggingCall(Level level, Marker marker, String messagePattern,
+        Object[] arguments, Throwable throwable) {
+
+      if (arguments == null) {
+        arguments = new Object[0];
+      }
+      if (!condition.apply(messagePattern, Arrays.asList(arguments))) {
+        delegate.atLevel(level).addMarker(marker).setCause(throwable).log(messagePattern,
+            arguments);
+      } else {
+        delegate.atLevel(elevatedLevel).addMarker(marker).setCause(throwable).log(messagePattern,
+            arguments);
+      }
+
+    }
+
+  }
+
+  /**
+   * A Logger implementation that will suppress duplicate messages within the supplied duration.
+   */
+  public static class DeduplicatingLogger extends ConditionalLogger {
+
+    private static final long serialVersionUID = 1L;
+
+    public DeduplicatingLogger(Logger log, Duration threshold) {
+      super(log, new BiFunction<>() {
+
+        private final Cache<String,List<Object>> cache =
```

Comment:
I don't think weakKeys() and weakValues() are what you would want to use here as it will evict if no more references. If you want to remove when low on memory you could look at using soft references instead.

But generally speaking though I think it's a better idea to simply put a maximum size on the cache and avoid weak/soft keys and references, there's more info about it here:  https://github.com/ben-manes/caffeine/wiki/Eviction#reference-based

Reply:
I removed the weak* method invocations and set the maximum size to 250 in 50de5fc. We can debate out making it a parameter or not.